src/zkpp.ts

```diff
@@ -0,0 +1,178 @@
+import { Buffer } from 'buffer'
+import scrypt, { type Options } from 'scrypt-async'
+import nacl from 'tweetnacl'
+import { AUTHSALT, CONTRACTSALT, CS, SALT_LENGTH_IN_OCTETS, SU } from './zkppConstants.js'
+
+// .toString('base64url') only works in Node.js
+export const base64ToBase64url = (s: string): string => s.replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')
+
+export const base64urlToBase64 = (s: string): string => s.replace(/_/g, '/').replace(/-/g, '+') + '='.repeat((4 - s.length % 4) % 4)
+
+export const hashStringArray = (...args: Array<Uint8Array | string>): Uint8Array => {
+  return nacl.hash(Buffer.concat(args.map((s) => nacl.hash(Buffer.from(s as string)))))
+}
+
+export const hashRawStringArray = (...args: Array<Uint8Array | string>): Uint8Array => {
+  return nacl.hash(Buffer.concat(args.map((s) => Buffer.from(s as string))))
+}
+
+export const randomNonce = (): string => {
+  return base64ToBase64url(Buffer.from(nacl.randomBytes(12)).toString('base64'))
+}
+
+export const hashRawB64url = (v: string | Buffer): string => {
+  return base64ToBase64url(Buffer.from(nacl.hash(Buffer.from(v as string))).toString('base64'))
+}
+
+export const hash = (v: string | Buffer): string => {
+  return base64ToBase64url(Buffer.from(nacl.hash(Buffer.from(v as string))).toString('base64'))
+}
+
+export const computeCAndHc = (r: string, s: string, h: string): [Uint8Array, Uint8Array] => {
+  const ħ = hashStringArray(r, s)
+  const c = hashStringArray(h, ħ)
+  const hc = nacl.hash(c)
+
+  return [c, hc]
+}
+
+export const encryptContractSalt = (c: Uint8Array, contractSalt: string): string => {
+  const encryptionKey = hashRawStringArray(CS, c).slice(0, nacl.secretbox.keyLength)
+  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
+
+  const encryptedContractSalt = nacl.secretbox(Buffer.from(contractSalt), nonce, encryptionKey)
+
+  return base64ToBase64url(Buffer.concat([nonce, encryptedContractSalt]).toString('base64'))
+}
+
+export const decryptContractSalt = (c: Uint8Array, encryptedContractSaltBox: string): string => {
+  const encryptionKey = hashRawStringArray(CS, c).slice(0, nacl.secretbox.keyLength)
+  const encryptedContractSaltBoxBuf = Buffer.from(base64urlToBase64(encryptedContractSaltBox), 'base64')
+
+  const nonce = encryptedContractSaltBoxBuf.subarray(0, nacl.secretbox.nonceLength)
+  const encryptedContractSalt = encryptedContractSaltBoxBuf.subarray(nacl.secretbox.nonceLength)
```

Comment:
Note: this changes the code slightly:

```diff
-  const nonce = encryptedContractSaltBoxBuf.slice(0, nacl.secretbox.nonceLength)
-  const encryptedContractSalt = encryptedContractSaltBoxBuf.slice(nacl.secretbox.nonceLength)
+  const nonce = encryptedContractSaltBoxBuf.subarray(0, nacl.secretbox.nonceLength)
+  const encryptedContractSalt = encryptedContractSaltBoxBuf.subarray(nacl.secretbox.nonceLength)
```

Did you intend to convert `.slice` to `.subarray`?

Reply:
I did, since `Buffer.prototype.slice` is deprecated.